File: app/src/main/java/ai/elimu/content_provider/room/dao/LetterDao.java
package ai.elimu.content_provider.room.dao;

import android.database.Cursor;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

import ai.elimu.content_provider.room.entity.Letter;

@Dao
public interface LetterDao {

    @Insert
    void insert(Letter letter);

    @Query("SELECT * FROM Letter WHERE id = :id")
    Letter load(Long id);

    @Query("SELECT * FROM Letter WHERE id = :id")
    Cursor loadAsCursor(Long id);

    @Query("SELECT * FROM Letter ORDER BY usageCount DESC")
    List<Letter> loadAllOrderedByUsageCount();

    @Query("SELECT * FROM Letter ORDER BY usageCount DESC")
    Cursor loadAllOrderedByUsageCount_Cursor();

    @Update
    void update(Letter letter);
}

File: app/src/main/java/ai/elimu/content_provider/room/dao/LetterDao.kt
package ai.elimu.content_provider.room.dao

import ai.elimu.content_provider.room.entity.Letter
import android.database.Cursor
import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update

@Dao
interface LetterDao {
    @Insert
    fun insert(letter: Letter)

    @Query("SELECT * FROM Letter WHERE id = :id")
    fun load(id: Long?): Letter?

    @Query("SELECT * FROM Letter WHERE id = :id")
    fun loadCursor(id: Long?): Cursor

    @Query("SELECT * FROM Letter ORDER BY usageCount DESC")
    fun loadAllOrderedByUsageCount(): MutableList<Letter>

    @Query("SELECT * FROM Letter ORDER BY usageCount DESC")
    fun loadAllOrderedByUsageCountCursor(): Cursor

    @Query("SELECT * FROM Letter l " +
            "INNER JOIN LetterSound_Letter lsl ON l.id = lsl.letters_id " +
            "WHERE lsl.LetterSound_id = :letterSoundId " +
            "ORDER BY lsl.letters_ORDER")
    fun loadAllByLetterSound(letterSoundId: Long?): Cursor

    @Update
    fun update(letter: Letter)

    @Query("DELETE FROM Letter")
    fun deleteAll()
}
